File: mastra_agent/src/runHandler.ts
import { type Request, type Response } from 'express';
import { randomUUID } from 'crypto';
import { buildAgent } from './agentFactory';
import { AGUIEmitter } from './aguiEmitter';
import { getConfig } from './config';
import { type ToolSchema, type RunCtx } from './bffToolAdapter';

function formatSse(event: Record<string, unknown>): string {
  return `data: ${JSON.stringify(event)}\n\n`;
}

export async function handleRun(req: Request, res: Response): Promise<void> {
  const body = (req.body ?? {}) as Record<string, unknown>;
  const threadId = (body.threadId as string | undefined) ?? `t_${randomUUID().slice(0, 8)}`;
  const runId = (body.runId as string | undefined) ?? `r_${randomUUID().slice(0, 8)}`;
  const messages = (body.messages as Array<{ role: string; content: string }> | undefined) ?? [];
  const rawTools = (body.tools as Array<Record<string, unknown>> | undefined) ?? [];
  const ctx = (body.context as Record<string, unknown> | undefined) ?? {};

  const toolSchemas: ToolSchema[] = rawTools.map((t) => ({
    name: t.name as string,
    description: (t.description as string | undefined) ?? '',
    inputSchema: (t.inputSchema as Record<string, unknown> | undefined) ?? { type: 'object', properties: {} },
  }));

  const sessionId = (ctx.sessionId as string | undefined) ?? '';
  const cfg = getConfig();
  const runCtx: RunCtx = {
    bffToolUrl: (ctx.bffToolUrl as string | undefined) || cfg.bffToolUrl,
    bffInternalSecret: cfg.bffInternalSecret,
    sessionId,
  };
  const model = (ctx.model as string | undefined) ?? cfg.model;

  res.setHeader('Content-Type', 'text/event-stream');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('X-Accel-Buffering', 'no');
  res.flushHeaders();

  const emitter = new AGUIEmitter(runId, threadId, async (event) => {
    res.write(formatSse(event));
  });

  try {
    await emitter.onRunStart();
    const agent = buildAgent(toolSchemas, runCtx, model);
    const userMessage =
      [...messages].reverse().find((m) => m.role === 'user')?.content ?? '';

    const stream = await agent.stream(userMessage);
    let streaming = false;

    // textStream is a ReadableStream<string>; Node 18+ supports async iteration on it
    for await (const chunk of stream.textStream as unknown as AsyncIterable<string>) {
      if (!streaming) {
        await emitter.onLlmStart();
        streaming = true;
      }
      await emitter.onLlmToken(chunk);
    }

    if (streaming) await emitter.onLlmEnd();
    await emitter.onRunEnd();
  } catch (err) {
    await emitter.onError(err instanceof Error ? err : new Error(String(err)));
  } finally {
    res.end();
  }
}

File: openai_agent/src/run_handler.py
"""POST /run — accepts BFF run payload, returns AG-UI SSE stream."""
from __future__ import annotations
import asyncio
import json
import logging
import uuid
from typing import AsyncGenerator

from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse
from agents import Runner

from .agent_factory import build_agent
from .agui_emitter import AGUIEmitter
from .config import get_config

logger = logging.getLogger(__name__)
router = APIRouter()


def _format_sse(event: dict) -> str:
    return f"data: {json.dumps(event)}\n\n"


@router.post("/run")
async def agent_run(request: Request) -> StreamingResponse:
    body = await request.json()
    thread_id: str = body.get("threadId", f"t_{uuid.uuid4().hex[:8]}")
    run_id: str = body.get("runId", f"r_{uuid.uuid4().hex[:8]}")
    messages: list = body.get("messages", [])
    tool_schemas: list = body.get("tools", [])
    ctx: dict = body.get("context", {})
    # Vertical persona forwarded by the BFF from the active vertical manifest.
    # Used to override the default banking system prompt so the agent replies
    # in the active vertical's language (care, retail, sports, workforce, etc.).
    vertical_flavor: str | None = body.get("vertical_flavor") or None

    bff_tool_url = ctx.get("bffToolUrl", "")
    session_id = ctx.get("sessionId", "")

    cfg = get_config()
    # Per-run overrides from BFF context win over server-side defaults. Empty
    # strings fall back to config so the operator can leave context.model unset.
    model = ctx.get("model") or cfg.model
    run_ctx = {
        "bff_tool_url": bff_tool_url or cfg.bff_tool_url,
        "bff_internal_secret": cfg.bff_internal_secret,
        "session_id": session_id,
        "base_url": cfg.llm_base_url,
    }

    return StreamingResponse(
        _stream(run_id, thread_id, messages, tool_schemas, run_ctx, model, cfg.llm_api_key, vertical_flavor),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


async def _stream(
    run_id: str,
    thread_id: str,
    messages: list,
    tool_schemas: list,
    run_ctx: dict,
    model: str,
    api_key: str,
    vertical_flavor: str | None = None,
) -> AsyncGenerator[str, None]:
    queue: asyncio.Queue = asyncio.Queue()

    async def sink(event: dict) -> None:
        await queue.put(event)

    emitter = AGUIEmitter(run_id=run_id, thread_id=thread_id, sink=sink)

    async def run_agent() -> None:
        try:
            await emitter.on_run_start()
            agent = build_agent(
                tool_schemas=tool_schemas,
                run_ctx=run_ctx,
                model=model,
                api_key=api_key,
                system_prompt=vertical_flavor,
            )
            user_input = next(
                (m["content"] for m in reversed(messages) if m.get("role") == "user"),
                "",
            )
            # Runner.run_streamed returns RunResultStreaming directly (not a context manager)
            result = Runner.run_streamed(agent, user_input)
            async for event in result.stream_events():
                await _handle_sdk_event(event, emitter)
            usage = getattr(result, "usage", None)
            if usage:
                await emitter.on_usage(
                    getattr(usage, "input_tokens", 0),
                    getattr(usage, "output_tokens", 0),
                )
            await emitter.on_run_end()
        except Exception as exc:
            logger.exception("[openai-agent] run error run=%s", run_id)
            await emitter.on_error(exc)
        finally:
            await queue.put(None)

    agent_task = asyncio.create_task(run_agent())

    try:
        while True:
            item = await queue.get()
            if item is None:
                break
            yield _format_sse(item)
    finally:
        agent_task.cancel()


async def _handle_sdk_event(event, emitter: AGUIEmitter) -> None:
    """Map openai-agents stream events to AG-UI emitter calls."""
    try:
        from agents.stream_events import RawResponsesStreamEvent, RunItemStreamEvent
        from openai.types.responses import ResponseTextDeltaEvent
    except ImportError:
        return

    if isinstance(event, RawResponsesStreamEvent):
        data = event.data
        if isinstance(data, ResponseTextDeltaEvent):
            if not emitter._current_message_id:
                await emitter.on_llm_start()
            await emitter.on_llm_token(data.delta)
    elif isinstance(event, RunItemStreamEvent):
        item = event.item
        item_type = getattr(item, "type", None)
        if item_type == "tool_call_item":
            raw = getattr(item, "raw_item", {})
            tc_id = raw.get("call_id", uuid.uuid4().hex[:12]) if isinstance(raw, dict) else uuid.uuid4().hex[:12]
            name = raw.get("name", "unknown") if isinstance(raw, dict) else "unknown"
            args = raw.get("arguments", "{}") if isinstance(raw, dict) else "{}"
            await emitter.on_tool_start(name, tc_id, args)
        elif item_type == "tool_call_output_item":
            raw = getattr(item, "raw_item", {})
            tc_id = raw.get("call_id", "") if isinstance(raw, dict) else ""
            output = getattr(item, "output", "")
            await emitter.on_tool_end(tc_id, output)
        elif item_type == "message_output_item":
            await emitter.on_llm_end()
